src/node.rs

```diff
@@ -411,520 +412,549 @@ impl PbftNode {
         let peers = get_peers_from_settings(&settings);
+        let new_peers_set: HashSet<PeerId> = peers.iter().cloned().collect();
 
-        let peer_ids: HashSet<_> = peers
-            .iter()
-            .cloned()
-            .filter(|pid| pid != &block.signer_id)
-            .collect();
-
-        if !voter_ids.is_subset(&peer_ids) {
-            return Err(PbftError::InternalError(format!(
-                "Got unexpected vote IDs: {:?}",
-                voter_ids.difference(&peer_ids).collect::<Vec<_>>()
-            )));
-        }
+        // Check if membership has changed
+        let old_peers_set: HashSet<PeerId> = state.peer_ids.iter().cloned().collect();
 
-        // Check that we've received 2f votes, since the primary vote is implicit
-        if voter_ids.len() < 2 * state.f as usize {
-            return Err(PbftError::InternalError(format!(
-                "Need {} votes, only found {}!",
-                2 * state.f,
-                voter_ids.len()
-            )));
+        if new_peers_set != old_peers_set {
+            state.peer_ids = peers;
+            let f = ((state.peer_ids.len() - 1) / 3) as u64;
+            if f == 0 {
+                panic!("This network no longer contains enough nodes to be fault tolerant");
+            }
+            state.f = f;
+            return true;
         }
 
-        Ok(())
+        false
     }
 
-    /// Use the given block's consensus seal to verify and commit the block this node is working on
-    fn catchup(&mut self, state: &mut PbftState, block: &Block) -> Result<(), PbftError> {
-        info!(
-            "{}: Trying catchup to #{} from BlockNew message #{}",
-            state, state.seq_num, block.block_num,
-        );
+    // ---------- Methods for building & verifying proofs and signed messages from other nodes ----------
 
-        match state.working_block {
-            Some(ref working_block) => {
-                let block_num_matches = block.block_num == working_block.get_block_num() + 1;
-                let block_id_matches = block.previous_id == working_block.get_block_id();
-
-                if !block_num_matches || !block_id_matches {
-                    error!(
-                        "Block didn't match for catchup: {:?} {:?}",
-                        block, working_block
-                    );
-                    return Err(PbftError::MismatchedBlocks(vec![
-                        pbft_block_from_block(block.clone()),
-                        working_block.clone(),
-                    ]));
-                }
-            }
-            None => {
-                error!(
-                    "Trying to catch up, but node does not have block #{} yet",
-                    state.seq_num
-                );
-                return Err(PbftError::NoWorkingBlock);
-            }
-        }
+    /// Generate a `protobuf::RepeatedField` of signed votes from a list of parsed messages
+    #[allow(clippy::needless_pass_by_value)]
```

Comment:
Couldn't we just pass in a slice here?

Reply:
Was copied as-is, fixed in new commit: "Remove needless_pass_by_value allows"